projects/storefrontlib/src/cms-components/misc/qualtrics/qualtrics-loader.service.ts

```diff
@@ -1,78 +1,131 @@
-import { QualtricsConfig } from './config/qualtrics-config';
+import { fromEvent, Observable, of } from 'rxjs';
+import { filter, map, switchMap, tap } from 'rxjs/operators';
 
+/**
+ * Service to integration Qualtrics.
+ *
+ * The integration observes the Qualtrics API, and when available, it runs the QSI API
+ * to let Qualtrics evaluate the application.
+ *
+ * The service supports an additional _hook_ (`isDataLoaded()`) that can be used to load application
+ * data before pulling the QSI API. This is benefitial in a single page application when additional
+ * data is required befor the Qualtrics _creatives_ run.
+ *
+ * This service also supports the creation of the Qualtrics deployment script. This is optional, as
+ * the script can be added in various ways.
+ */
 @Injectable({
   providedIn: 'root',
 })
 export class QualtricsLoaderService {
-  private qualtricsLoaded$ = new BehaviorSubject<boolean>(false);
+  /**
+   * Reference to the QSI API.
+   */
+  protected qsiApi: any;
 
-  constructor(private winRef: WindowRef, private config: QualtricsConfig) {
-    if (
-      Boolean(this.winRef.nativeWindow) &&
-      Boolean(this.winRef.document) &&
-      this.isQualtricsConfigured()
-    ) {
-      this.initialize();
-      this.setup();
-    }
-  }
+  /**
+   * QSI load event that happens when the QSI JS file is loaded.
+   */
+  private qsiLoaded$: Observable<any> = this.winRef?.nativeWindow
+    ? fromEvent(this.winRef.nativeWindow, 'qsi_js_loaded')
+    : of();
 
-  private initialize(): void {
-    fromEvent(this.winRef.nativeWindow, 'qsi_js_loaded').subscribe(() =>
-      this.qualtricsLoaded$.next(true)
-    );
-  }
+  /**
+   * Emits the Qualtrics Site Intercept (QSI) JavaScript API whenever available.
+   *
+   * The API is emitted when the JavaScript resource holding this API is fully loaded.
+   * The API is also stored locally in the service, in case it's required later on.
+   */
+  protected qsi$: Observable<any> = this.qsiLoaded$.pipe(
+    switchMap(() => this.isDataLoaded()),
+    map(() => this.winRef?.nativeWindow['QSI']),
+    filter((api) => Boolean(api)),
+    tap((qsi) => (this.qsiApi = qsi))
+  );
 
-  private setup(): void {
-    const qualtricsScript = this.winRef.document.createElement('script');
-    qualtricsScript.type = 'text/javascript';
-    qualtricsScript.defer = true;
-    qualtricsScript.src = 'assets/qualtricsIntegration.js';
+  constructor(
+    protected winRef: WindowRef,
+    protected rendererFactory: RendererFactory2
+  ) {
+    this.initialize();
+  }
 
-    const idScript = this.winRef.document.createElement('div');
-    idScript.id = this.config.qualtrics.projectId;
+  /**
+   * Starts observing the Qualtrics integration. The integration is based on a
+   * Qualtrics specific event (`qsi_js_loaded`). As soon as this events happens,
+   * we run the API.
+   */
+  protected initialize() {
+    this.qsi$.subscribe(() => this.run());
```

Comment:
@bgambocjaviniar pointed out that `run()` is called twice in a setup where a CMS component is used and the script is manually added to `index.html`:
- when CMS component is instantiated, the `addScript` is called, which then calls the `run()`
- when the service is instantiated, the `run` is called from this spot

Reply:
correct, that would be the consequences in the current implementation. We could avoid this by:
1. tell the customer to either drop the global added script, or not use the cms component; we should understand what's the point of this mixed approach. I believe this would not be a recommendation anyway. 
2. not adding a global configuration for the script (I'd actually be happy to drop the default 'assets/qualtricsIntegration.js'). 
3. introduce an additional option in the config to tell the component to not add the script. but it's really going in an unclear direction in my pov.

Reply:
1. I like this option, as it still offers various options for customers. It should really need be either the component approach, or manually including the script in the `index.html`.
2. You mean to not expect any default name, and require a configuration input from customers (only if they are not using the CMS component approach)? 
3. Agreed, this will also bring a bit more complexity into this integration.

Reply:
I prefer to dop the default configuration for the script location (`qualtrics.scriptSource: 'assets/qualtricsIntegration.js'`). If the customer adds the script manually inside the index (or any other place), he will not be bothered with this default in case he'd have sampledata with this component.
In addition to this change, we should be focal (`console.log`) about the lack of a global config in case it's used. I'd do this in `devMode()` only.

Reply:
Sure, let's drop it, and log the lack of a global config.